server/util/util.ts

```diff
@@ -112,3 +112,65 @@ export const timestampColumnTransformer: ValueTransformer = {
 };
 
 export const oneDaySeconds = 24 * 60 * 60;
+
+export type TimeInterval =
+  | 'day'
+  | 'week'
+  | 'month'
+  | 'year';
+
+export function getEsTimeInterval(interval: TimeInterval) {
+  switch (interval) {
+    case 'day': return 'd';
+    case 'week': return 'w';
+    case 'month': return 'M';
+    case 'year': return 'y';
+    default:
+      throw new Error(`Unsupported time interval '${interval}'`);
+  }
+}
+
+export function getEsDateFormat(interval: TimeInterval) {
+  switch (interval) {
+    case 'day':
+    case 'week': return 'yyyy-MM-dd';
+    case 'month': return 'yyyy-MM';
+    case 'year': return 'yyyy';
+    default:
+      throw new Error(`Unsupported interval '${interval}'`);
+  }
+}
+
+export function getMomentDateFormat(interval: TimeInterval) {
+  switch (interval) {
+    case 'day':
+    case 'week': return 'yyyy-MM-DD';
+    case 'month': return 'yyyy-MM';
+    case 'year': return 'yyyy';
+    default:
+      throw new Error(`Unsupported interval '${interval}'`);
+  }
+}
+
+export function requireQuery<TQuery extends object>(
+  req: ApiRequest<unknown, unknown, TQuery>,
+  keys: Array<keyof TQuery>,
+) {
+  const missing = getMissingKeys(req.query, keys);
+  if (missing.length) {
+    throw new BadRequestError(`Missing required query params: ${missing.join(', ')}`);
+  }
+
+  return req.query;
+}
+
+function getMissingKeys<T extends object>(obj: T, keys: Array<keyof T>) {
+  const missing = [];
```

Comment:
Purely preference here but - I find this to be more readable.
`return keys.filter(key => obj[key] === undefined);

Reply:
Yep, filter is much better. It totally slipped my mind when writing this.